### PR TITLE
refactor: moved metrics to single call

### PR DESCRIPTION
**Pull request by 0xcadams**
## Description

Moved metrics addition to be a single function which handles fetching the reporting URL internally.

### CODE REVIEW

**Comment by 0xcadams** [comment]
@victorges I think expecting users to destroy/recreate on source change is reasonable - see [Mux](https://docs.mux.com/guides/data/monitor-html5-video-element#new-source).

File: examples/svelte/src/lib/Video.svelte [hunk 1]
```diff
@@ -25,6 +25,10 @@
 
     const hlsSrc = getMediaSourceType(source);
 
+    addMediaMetrics(video, source, (e) =>
```
Comment (victorges) [inline examples/svelte/src/lib/Video.svelte:27]: Neat!

File: packages/core/src/media/index.ts [hunk 2]
```diff
@@ -1,10 +1,10 @@
 export { canPlayMediaNatively } from './browser';
 export { parseCid } from './decentralizedStorage';
 export {
+  addMediaMetrics,
   getMetricsReportingUrl,
```
Comment (victorges) [inline packages/core/src/media/index.ts]: Does it still make sense to export the `getMetricsReportingUrl`? Can it still be used for anything now that `reportMediaMetrics` is not exported?

File: packages/core/src/media/metrics/metrics.ts [hunk 3]
```diff
@@ -265,28 +270,34 @@ const VIDEO_METRICS_INITIALIZED_ATTRIBUTE = 'data-metrics-initialized';
...
+ * report them to a websocket. Automatically handles a redirect to get the
+ * metrics endpoint.
+ *
+ * @param element Element to capture playback metrics from
  */
-export function reportMediaMetrics<TElement extends HTMLMediaElement>(
-  element: TElement,
-  reportingWebsocketUrl: string,
+export function addMediaMetrics<TElement extends HTMLMediaElement>(
+  element: TElement | undefined | null,
```
Comment (victorges) [inline packages/core/src/media/metrics/metrics.ts:284]: WDYT of allowing receiving an HTML ID here as well like mux? Then one could do like `addMediaMetrics('#my-player', src)` (tho one will likely already have a ref to the element anyway to grab or set the `src` 🤷‍♂️)

File: packages/core/src/media/metrics/metrics.ts [hunk 4]
```diff
@@ -265,28 +270,34 @@ const VIDEO_METRICS_INITIALIZED_ATTRIBUTE = 'data-metrics-initialized';
...
+  sourceUrl: string | undefined | null,
+  onError?: (error: unknown) => void,
 ): MediaMetrics<TElement> {
   const defaultResponse: MediaMetrics<TElement> = {
     metrics: null,
-    websocket: null,
-    report: null,
+    destroy: () => {
+      //
```
Comment (victorges) [inline packages/core/src/media/metrics/metrics.ts:291]: What's the empty comment for? Is it some dark TS sorcery?

File: packages/core/src/media/metrics/metrics.ts [hunk 5]
```diff
@@ -265,28 +270,34 @@ const VIDEO_METRICS_INITIALIZED_ATTRIBUTE = 'data-metrics-initialized';
...
 
+  if (!element || !sourceUrl) {
+    return defaultResponse;
+  }
+
   // do not attach twice (to the same websocket)
-  if (element.getAttribute(VIDEO_METRICS_INITIALIZED_ATTRIBUTE) === 'true') {
+  if (element.hasAttribute(VIDEO_METRICS_INITIALIZED_ATTRIBUTE)) {
     return defaultResponse;
   }
```
Comment (victorges) [inline packages/core/src/media/metrics/metrics.ts:302]: Should this actually throw an error instead, or at least log? Or is there a valid use case for calling it multiple times?

File: packages/core/src/media/metrics/metrics.ts [hunk 6]
```diff
@@ -304,31 +315,52 @@ export function reportMediaMetrics<TElement extends HTMLMediaElement>(
...
-
-        // auto-reconnect with exponential backoff
-        setTimeout(function () {
-          createNewWebSocket(numRetries++);
-        }, Math.pow(2, numRetries) * 1e3);
-      };
+    const createNewWebSocket = async (numRetries = 0) => {
+      const reportingWebsocketUrl = await getMetricsReportingUrl(sourceUrl);
+
+      console.log(reportingWebsocketUrl);
```
Comment (victorges) [inline packages/core/src/media/metrics/metrics.ts]: Forgotten log?

**Review by victorges — APPROVED** [review]
Great!

File: examples/svelte/src/lib/Video.svelte (hunk 1 above)
Comment (0xcadams) [reply examples/svelte/src/lib/Video.svelte:27]: 😄

File: packages/core/src/media/index.ts (hunk 2 above)
Comment (0xcadams) [reply packages/core/src/media/index.ts]: Great point! No, I don't think it makes sense anymore. Removing

File: packages/core/src/media/metrics/metrics.ts (hunk 3 above)
Comment (0xcadams) [reply packages/core/src/media/metrics/metrics.ts:284]: I don't love that API - it feels very JQuery to me and I don't think it's very common anymore. A user could easily do something like:

```tsx
const videoElement = document.querySelector('#videoPlayer');

if (videoElement instanceof HTMLMediaElement) {
  addMediaMetrics(videoElement);
}
```

And conform to this API. I think as a library, we should push best practices, and getting direct refs is much less error prone than using HTML selectors.

File: packages/core/src/media/metrics/metrics.ts (hunk 4 above)
Comment (0xcadams) [reply packages/core/src/media/metrics/metrics.ts:291]: Haha it's a lint hack so it doesn't complain 🙃

File: packages/core/src/media/metrics/metrics.ts (hunk 5 above)
Comment (0xcadams) [reply packages/core/src/media/metrics/metrics.ts:302]: No I don't think this should be an error, it's fine if the component tries to initialize this multiple times

File: packages/core/src/media/metrics/metrics.ts (hunk 6 above)
Comment (0xcadams) [reply packages/core/src/media/metrics/metrics.ts]: 😵‍💫

File: packages/core/src/media/metrics/metrics.ts (hunk 3 above)
Comment (victorges) [reply packages/core/src/media/metrics/metrics.ts:284]: Cool that makes a lot of sense.

**Pull request merged**